### ISSUE TITLE
source_catalog failing for some MIRI and NIRCam images

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#4913]
_Issue [JP-1440](https://jira.stsci.edu/browse/JP-1440) was created on JIRA by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse):_

DMS reports an error being thrown by the calwebb_image3 source_catalog step for some MIRI and NIRCam datasets using the latest 0.16.0 (B7.5rc1) delivery in their test environment. Traceback from a MIRI example is:
```java
2020-05-08 05:12:39,792 - CRDS - DEBUG - getreferences() CRDS version: 7.4.1.3, b7.4.1.3, b38c37e3437101250cbf746e1acc2fbfb34a8005 
2020-05-08 05:12:39,792 - CRDS - DEBUG - getreferences() server: https://jwst-serverless-mode.stsci.edu 
2020-05-08 05:12:39,792 - CRDS - DEBUG - getreferences() observatory: jwst 
2020-05-08 05:12:39,792 - CRDS - DEBUG - getreferences() reftypes: ('abvegaoffset',) 
2020-05-08 05:12:39,792 - CRDS - DEBUG - getreferences() context: None 
2020-05-08 05:12:39,792 - CRDS - DEBUG - getreferences() ignore_cache: False 
2020-05-08 05:12:39,792 - CRDS - DEBUG - CRDS_PATH = ████████████████ 
2020-05-08 05:12:39,792 - CRDS - DEBUG - CRDS_SERVER_URL = https://jwst-serverless-mode.stsci.edu 
2020-05-08 05:12:39,793 - CRDS - DEBUG - Final effective context is 'jwst_0614.pmap' 
2020-05-08 05:12:39,793 - CRDS - DEBUG - Computing best references locally. 
2020-05-08 05:12:39,794 - CRDS - DEBUG - Bestrefs header: 
2020-05-08 05:12:39,794 - CRDS - DEBUG - {'META.EXPOSURE.READPATT [READPATT]': 'FAST', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.EXPOSURE.TYPE [EXP_TYPE]': 'MIR_IMAGE', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.INSTRUMENT.BAND [BAND]': 'UNDEFINED', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.INSTRUMENT.CHANNEL [CHANNEL]': 'UNDEFINED', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.INSTRUMENT.CORONAGRAPH [CORONMSK]': 'UNDEFINED', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.INSTRUMENT.DETECTOR [DETECTOR]': 'MIRIMAGE', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.INSTRUMENT.FILTER [FILTER]': 'F770W', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.INSTRUMENT.LAMP_STATE [LAMP]': 'OFF', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.INSTRUMENT.NAME [INSTRUME]': 'MIRI', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.OBSERVATION.DATE [DATE-OBS]': '2016-01-17', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.OBSERVATION.TIME [TIME-OBS]': '01:04:27.879000', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.SUBARRAY.NAME [SUBARRAY]': 'SUB128', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'META.VISIT.TSOVISIT [TSOVISIT]': 'F', 
2020-05-08 05:12:39,794 - CRDS - DEBUG - 'REFTYPE': 'UNDEFINED'} 
2020-05-08 05:12:39,795 - CRDS - DEBUG - Reference type 'abvegaoffset' defined as 'jwst_miri_abvegaoffset_0001.asdf' 
2020-05-08 05:12:39,795 - stpipe.Image3Pipeline.source_catalog - INFO - Using ABVEGAOFFSET reference file ████████████████████████████████████████████████████████████████ 
2020-05-08 05:12:39,795 - stpipe.Image3Pipeline.source_catalog - INFO - Instrument: MIRI 
2020-05-08 05:12:39,795 - stpipe.Image3Pipeline.source_catalog - INFO - Detector: MIRIMAGE 
2020-05-08 05:12:39,795 - stpipe.Image3Pipeline.source_catalog - INFO - Filter: F770W 
2020-05-08 05:12:39,796 - stpipe.Image3Pipeline.source_catalog - INFO - Subarray: SUB128 
Traceback (most recent call last): 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/bin/strun", line 26, in <module> 
step = Step.from_cmdline(sys.argv[1:]) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/step.py", line 156, in from_cmdline 
return cmdline.step_from_cmdline(args) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/cmdline.py", line 346, in step_from_cmdline 
step.run(*positional) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/step.py", line 392, in run 
step_result = self.process(*args) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/pipeline/calwebb_image3.py", line 111, in process 
self.source_catalog(result) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/step.py", line 392, in run 
step_result = self.process(*args) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/source_catalog/source_catalog_step.py", line 76, in process 
model.data -= bkg.background 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/astropy/utils/decorators.py", line 758, in __get__ 
val = self.fget(obj) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/source_catalog/source_catalog.py", line 334, in background 
return self._background2d.background 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/astropy/utils/decorators.py", line 758, in __get__ 
val = self.fget(obj) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/source_catalog/source_catalog.py", line 321, in _background2d 
sigma_clip=sigma_clip, bkg_estimator=bkg_estimator) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/photutils/background/background_2d.py", line 338, in __init__ 
self._calc_bkg_bkgrms() 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/photutils/background/background_2d.py", line 671, in _calc_bkg_bkgrms 
idx = self._select_meshes(data_sigclip) 
File "/ifs/pipeline/dms/rhel7/pkgs/miniconda3-jwst/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/photutils/background/background_2d.py", line 455, in _select_meshes 
self.exclude_percentile)) 
ValueError: All meshes contain > 1000.0 (10.0 percent per mesh) masked pixels. Please check your data or decrease "exclude_percentile". 
---------------------------------------------------------------------- 
ERROR RUNNING STEP 'Image3Pipeline': 
All meshes contain > 1000.0 (10.0 percent per mesh) masked pixels. 
Please check your data or decrease "exclude_percentile". 
----------------------------------------------------------------------
```

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1440?focusedCommentId=463052#comment-463052):_

Took a quick look at the MIRI example. It consists of a single exposure using the SUB128 subarray, which is a small 128x128 subarray in the upper left corner of the detector coincident with the Lyot mask. The cal product SCI array shows nothing but semi-random noise (no sources), and the DQ array is loaded with 4's (JUMP_DET), indicating that many pixels were flagged as containing jumps in the detector1 pipeline, but they are NOT flagged with DO_NOT_USE. This is consistent with the fact that the i2d (resampled) product shows non-zero values for nearly all pixels in the CON and WHT arrays, which I assume means those pixels are NOT masked out when used within source_catalog.

So it will have to be up to the expert - [Larry Bradley](https://jira.stsci.edu/secure/ViewProfile.jspa?name=lbradley) - do determine the cause of the error message. Note that because there aren't any obvious real sources in the image, it may be trying to work with bogus detections.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1440?focusedCommentId=463413#comment-463413):_

A slightly different kind of source_catalog error has been seen when processing NIRISS images. The traceback for this is:
```java
Traceback (most recent call last):
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/bin/strun", line 26, in <module>
    step = Step.from_cmdline(sys.argv[1:])
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/step.py", line 156, in from_cmdline
    return cmdline.step_from_cmdline(args)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/cmdline.py", line 346, in step_from_cmdline
    step.run(*positional)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/step.py", line 392, in run
    step_result = self.process(*args)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/pipeline/calwebb_image3.py", line 111, in process
    self.source_catalog(result)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/step.py", line 392, in run
    step_result = self.process(*args)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/source_catalog/source_catalog_step.py", line 76, in process
    model.data -= bkg.background
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/astropy/utils/decorators.py", line 758, in __get__
    val = self.fget(obj)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/source_catalog/source_catalog.py", line 334, in background
    return self._background2d.background
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/astropy/utils/decorators.py", line 758, in __get__
    val = self.fget(obj)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/source_catalog/source_catalog.py", line 321, in _background2d
    sigma_clip=sigma_clip, bkg_estimator=bkg_estimator)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/photutils/background/background_2d.py", line 338, in __init__
    self._calc_bkg_bkgrms()
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/photutils/background/background_2d.py", line 691, in _calc_bkg_bkgrms
    bkg = self._interpolate_meshes(self._bkg1d)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/photutils/background/background_2d.py", line 585, in _interpolate_meshes
    interp_func = ShepardIDWInterpolator(yx, data)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/photutils/utils/interpolation.py", line 136, in __init__
    raise ValueError('The number of values must match the number '
ValueError: The number of values must match the number of coordinates.
----------------------------------------------------------------------
ERROR RUNNING STEP 'Image3Pipeline':
    The number of values must match the number of coordinates.
----------------------------------------------------------------------
```
An example dataset has been copied to the same data directory at ███████████████████████████████████████████ It is image3 ASN "jw87700-o009_20200511t140515_image3_001_asn.json" containing 3 input cal products (all of which have been copied).

**Comment by stscijgbot-jp** [comment]
_Comment by [Larry Bradley](https://jira.stsci.edu/secure/ViewProfile.jspa?name=lbradley) on [JIRA](https://jira.stsci.edu/browse/JP-1440?focusedCommentId=463434#comment-463434):_

#4929 also fixes the issue with NIRISS data.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1440?focusedCommentId=463445#comment-463445):_

One more slightly different error signature seen when processing ASN "jw00793-o024_20200511t154524_image3_001_asn.json", which has been copied to the ███████████████████████████████████████████ Can you check this one too, to see if it's fixed by the same update?
```java
Traceback (most recent call last):
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/bin/strun", line 26, in <module>
    step = Step.from_cmdline(sys.argv[1:])
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/step.py", line 156, in from_cmdline
    return cmdline.step_from_cmdline(args)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/cmdline.py", line 346, in step_from_cmdline
    step.run(*positional)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/step.py", line 392, in run
    step_result = self.process(*args)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/pipeline/calwebb_image3.py", line 111, in process
    self.source_catalog(result)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/step.py", line 392, in run
    step_result = self.process(*args)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/source_catalog/source_catalog_step.py", line 99, in process
    catalog = catobj.catalog
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/astropy/utils/decorators.py", line 758, in __get__
    val = self.fget(obj)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/source_catalog/source_catalog.py", line 1297, in catalog
    catalog[colname] = getattr(self, column)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/astropy/utils/decorators.py", line 758, in __get__
    val = self.fget(obj)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/source_catalog/source_catalog.py", line 1147, in nn_abmag
    nn_abmag = self.aper_total_abmag[self._ckdtree_query[1]]
IndexError: index 1 is out of bounds for axis 0 with size 1
----------------------------------------------------------------------
ERROR RUNNING STEP 'Image3Pipeline':
    index 1 is out of bounds for axis 0 with size 1
----------------------------------------------------------------------
```

**Comment by stscijgbot-jp** [comment]
_Comment by [Larry Bradley](https://jira.stsci.edu/secure/ViewProfile.jspa?name=lbradley) on [JIRA](https://jira.stsci.edu/browse/JP-1440?focusedCommentId=463450#comment-463450):_

This issue is unrelated to background subtraction and is not fixed by my open PR.  This issue is triggered by detecting exactly one source – the nearest-neighbor calculation fails to take that into account.  I'll open a different PR to fix it.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1440?focusedCommentId=463631#comment-463631):_

All 3 flavors of errors fixed by #4929
